kernel: split k_busy_wait() out of timeout.c

npitre opened the pull request:
This allows for builds with `CONFIG_SYS_CLOCK_EXISTS=n` in which case
busy waits are achieved with a crude CPU loop. If ever accuracy is
needed even with such a configuration then implementing `arch_busy_wait()`
should be considered. Otherwise the code is unchanged.

npitre committed:
kernel: split k_busy_wait() out of timeout.c

This will allow for builds with CONFIG_SYS_CLOCK_EXISTS=n. For now this
is only the code move.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

peter-mitsis commented on kernel/busy_wait.c:
```diff
@@ -0,0 +1,66 @@
+	}
+#else
+	/*
+	 * Crud busy loop assuming 1 GHz CPU and 2 cycles per loop.
```
I am guessing that we would prefer to use "Crude" instead "Crud" here.

npitre replied:
Indeed, fixed.

peter-mitsis commented on kernel/busy_wait.c:
```diff
@@ -0,0 +1,66 @@
+
+	do {
+		arch_nop();
+	} while (loops-- > 0);
```
Two items related to this loop.

1. The combination of both the do-while loop and the post-decrement on _loops_ means that we are likely to go through the loop one more time than intended. I realise that this is a meant to be a crude approximation, so it may not be the most significant of observations.
2. Is it worthwhile to just have a Kconfig value to indicate the number of loops per microsecond? This could serve as a convenient manner to tune this delay until the DT method you described is implemented.

npitre replied:
1) The timing won't be accurate enough to matter, but I fixed the logic anyway.
2) Good idea, implemented.

andyross reviewed (commented):
Could we get the code motion split out from the addition of the loop count feature as separate patches?

Also, note that on nRF at least, ARCH_HAS_CUSTOM_BUSY_WAIT is in fact a CPU counter (though a tuned assembly one IIRC).  It might be nice to unify the kconfig/preprocessor selection of these somehow (e.g. have this new thing become a "CUSTOM_BUSY_WAIT" or the like).

npitre committed:
kernel: crude k_busy_wait() implementation

This allows for builds with CONFIG_SYS_CLOCK_EXISTS=n in which case
busy waits are achieved with a crude CPU loop. If ever accuracy is
needed even with such a configuration then implementing arch_busy_wait()
should be considered.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
> Could we get the code motion split out from the addition of the loop
> count feature as separate patches?

Good point, done.

I'm not clear about your "CUSTOM_BUSY_WAIT" remark though.

andyross commented:
Just a vague sense that if we have a platform with a "custom busy wait" that is actually a custom CPU-counted loop, and we now have an optional CPU-counted loop, that the nRF busy wait implementation might properly be seen as a special case of "cpu counted busy wait".

npitre commented:
> Just a vague sense that if we have a platform with a "custom busy wait" 
> that is actually a custom CPU-counted loop, and we now have an optional
> CPU-counted loop, that the nRF busy wait implementation might properly 
> be seen as a special case of "cpu counted busy wait".

Hmmmpfff. Is this important?

The loop I added is really about making the code build and work when system 
timers are configured out (e.g. some UART drivers use `k_busy_loop()` when 
polling status bits). It is written in C and doesn't pretend to be accurate 
at all (as long as it waits more than the time asked... in practice way 
more). It is absolutely not substituting for carefully crafted arch 
provided implementation. But when you disables timer support this usually 
means you don't really care about accurate time in the first place.

andyross commented:
Not important, no.  Just a design sense.  We now have two variant cpu-counted loops but for different purposes and configured differently.  A smell thing.

cfriedt reviewed (approved):
There are a couple of use cases for this.
